lib/faster_s3_url/builder.rb

```diff
@@ -103,34 +104,24 @@ def presigned_url(key, time: nil, expires_in: DEFAULT_EXPIRES_IN,
 
       credential_scope = datestamp + '/' + region + '/' + SERVICE + '/' + 'aws4_request'
 
-      canonical_query_string_parts = [
-          "X-Amz-Algorithm=#{ALGORITHM}",
-          "X-Amz-Credential=" + uri_escape(@access_key_id + "/" + credential_scope),
-          "X-Amz-Date=" + amz_date,
-          "X-Amz-Expires=" + expires_in.to_s,
-          "X-Amz-SignedHeaders=" + SIGNED_HEADERS,
-        ]
-
-      extra_params = {
-        :"response-cache-control" => response_cache_control,
-        :"response-content-disposition" => response_content_disposition,
-        :"response-content-encoding" => response_content_encoding,
-        :"response-content-language" => response_content_language,
-        :"response-content-type" => response_content_type,
-        :"response-expires" => convert_for_timestamp_shape(response_expires),
-        :"versionId" => version_id
+      # These have to be sorted, but sort is case-sensitive, and we have a fixed
+      # list of headers we know might be here... turns out they are already sorted?
+      canonical_query_params = {
+        'X-Amz-Algorithm': ALGORITHM,
+        "X-Amz-Credential": uri_escape(@access_key_id + "/" + credential_scope),
+        "X-Amz-Date": amz_date,
+        "X-Amz-Expires": expires_in.to_s,
+        "X-Amz-Security-Token": uri_escape(session_token),
+        "X-Amz-SignedHeaders": SIGNED_HEADERS,
+        "response-cache-control": uri_escape(response_cache_control),
+        "response-content-disposition": uri_escape(response_content_disposition),
+        "response-content-encoding": uri_escape(response_content_encoding),
+        "response-content-language": uri_escape(response_content_language),
+        "response-content-type": uri_escape(response_content_type),
+        "response-expires": uri_escape(convert_for_timestamp_shape(response_expires)),
+        "versionId": uri_escape(version_id)
       }.compact
-
-
-      if extra_params.size > 0
-        # These have to be sorted, but sort is case-sensitive, and we have a fixed
-        # list of headers we know might be here... turns out they are already sorted?
-        extra_param_parts = extra_params.collect {|k, v| "#{k}=#{uri_escape v}" }.join("&")
-        canonical_query_string_parts << extra_param_parts
-      end
-
-      canonical_query_string = canonical_query_string_parts.join("&")
-
+      canonical_query_string = canonical_query_params.collect {|k, v| "#{k}=#{v}" }.join("&")
```

Comment:
Loks like you removed the `if extra_params.size > 0` check, was that intentional?  I'm not sure if it matters, I guess it was supposed to increase performance in the 0 element case, but it may not actually, is that what you were thinking?

Reply:
That is what I was thinking.

I also needed to conditionally add `X-Amz-Security-Token` and the `.compact` call is removing all the nils we don't need for us, so the logic felt simpler to me.

The performance result seemed more or less the same to me.

Reply:
OK sounds good